File: apps/web/src/rag/retrieval.ts
import { searchSimilarChunks } from "../db.js";
import type { RelevantChunk } from "./types.js";

/**
 * Retrieve the most relevant knowledge chunks for a user's query.
 *
 * Uses PostgreSQL full-text search (tsvector/tsquery) to find
 * chunks whose content matches the query terms. This replaces
 * vector embeddings with built-in PostgreSQL text search —
 * no external API needed.
 */
export async function retrieveRelevantChunks(
  query: string,
  topK: number = 5,
  minSimilarity: number = 0
): Promise<RelevantChunk[]> {
  const results = await searchSimilarChunks(query, topK);
  return results.filter(chunk => chunk.similarity >= minSimilarity) as RelevantChunk[];
}

File: apps/api/src/api/rag/retrieval.py
"""Retrieval step of RAG — PostgreSQL full-text search over knowledge chunks."""
from __future__ import annotations

from . import db


async def retrieve_relevant_chunks(
    query: str, top_k: int = 5, min_similarity: float = 0.0
) -> list[dict]:
    results = await db.search_similar_chunks(query, top_k)
    return [c for c in results if c["similarity"] >= min_similarity]
